python/ray/spark/__init__.py

```diff
@@ -0,0 +1,593 @@
+        *_convert_ray_node_options(head_options),
+    ]
+
+    _logger.info(f"Start Ray head, command: {' '.join(ray_head_node_cmd)}")
+
+    with open(
+        os.path.join(ray_log_dir, "ray-start-head.log"), "w", buffering=1
+    ) as head_log_fp:
+        ray_head_proc = exec_cmd(
+            ray_head_node_cmd,
+            synchronous=False,
+            capture_output=False,
+            stream_output=False,
+            stdout=head_log_fp,
+            stderr=subprocess.STDOUT,
+        )
+
+    # wait ray head node spin up.
+    wait_ray_node_available(
+        ray_head_hostname,
+        ray_head_port,
+        40,
+        error_on_failure="Start Ray head node failed!",
+    )
+
+    _logger.info("Ray head node started.")
+
+    # NB:
+    # In order to start ray worker nodes on spark cluster worker machines,
+    # We launch a background spark job:
+    #  1. it is a barrier mode spark job, i.e. all spark tasks in the job run concurrently.
+    #     If the spark cluster resources are not sufficient to launch all tasks concurrently,
+    #     the spark job will hang and retry. If spark exceeds the maximum retry count, the
+    #     submitted operation will fail.
+    #  2. Each spark task launches one ray worker node. This design ensures all ray worker nodes
+    #     have the same shape (same cpus / gpus / memory configuration). If ray worker nodes have a
+    #     non-uniform shape, the Ray cluster setup will be non-deterministic and could create
+    #     issues with node sizing.
+    #  3. A ray worker node is started via the `ray start` CLI. In each spark task, a child
+    #     process is started and will execute a `ray start ...` command in blocking mode.
+    #  4. When the ray cluster is shutdown, killing ray worker nodes is implemented by:
+    #     Installing a PR_SET_PDEATHSIG signal for the `ray start ...` child processes
+    #     so that when parent process (pyspark task) is killed, the child processes
+    #     (`ray start ...` processes) will receive a SIGTERM signal, killing it.
+    #     Shutting down the ray cluster is performed by calling `sparkContext.cancelJobGroup`
+    #     to cancel the background spark job, sending a SIGKILL signal to all spark tasks.
+    #     Once the spark tasks are killed, this triggers the sending of a SIGTERM to the child
+    #     processes spawned by the `ray_start ...` process.
+
+    def ray_cluster_job_mapper(_):
+        from pyspark.taskcontext import TaskContext
+
+        _worker_logger = logging.getLogger("ray.spark.worker")
+
+        context = TaskContext.get()
+        task_id = context.partitionId()
+
+        # NB: If we launch multiple ray worker nodes at the same time,
+        #  it might cause Raylet to have a port conflict, likely due to a race condition.
+        #  A sleep is added here to attempt to avoid resource allocation contention with available
+        #  ports.
+        time.sleep(task_id + 5.0)
```

Comment:
2 issues:

* The interval "1 second" is too small , we need to increase to 10s interval between each worker starting time. Otherwise if many tasks running (e.g. 8 tasks on a node) port conflicts might happen frequently.

* Only spark tasks on a local machine need to set different delay time. e.g., there're 8 tasks, tasks 1-4 in nodeA, other tasks on nodeB, we hope:
task1-4: sleep 0s / 10s / 20s / 30s
task5-8: sleep 0s / 10s / 20s / 30s

To achieve this, because we cannot use `get_spark_task_local_rank` now, we can use exclusive file lock instead:

In each spark task, we first try to acquire an exclusive file lock, and keep locking it for 10s, and then release the file lock.

Reply:
Implemented a process-level flock that (should) eliminate race conditions by using file truncation and not attempting to remove the created lock file in the local /tmp/ directory. Let me know if you were thinking of something different here.

Reply:
@BenWilson2 
see this https://github.com/ray-project/ray/pull/28771#discussion_r1005174118

Reply:
changed the implementation :)